### PR TITLE
demux_mkv: support ARIB captions

### PR DESCRIPTION
**Pull request by rcombs**
As per the newly-merged spec. Nothing too complex here, though it does need an `AVCodecParameters` to set the profile + some wrapping to make sure it gets freed.

### CODE REVIEW

File: demux/demux_mkv.c [hunk 1]
```diff
@@ -1831,9 +1831,16 @@ static const char *const mkv_sub_tag[][2] = {
     { "D_WEBVTT/CAPTIONS",  "webvtt-webm"},
     { "S_TEXT/WEBVTT",      "webvtt"},
     { "S_DVBSUB",           "dvb_subtitle"},
+    { "S_ARIBSUB",          "arib_caption"},
     {0}
 };
 
+static void avcodec_par_destructor(void *p)
+{
+    if (p)
```
Comment (jeeb) [inline demux/demux_mkv.c:1840]: Destructor only gets registered when after the talloc, so probably this nullptr check is not required.

File: demux/demux_mkv.c [hunk 2]
```diff
@@ -1862,6 +1869,45 @@ static int demux_mkv_open_sub(demuxer_t *demuxer, mkv_track_t *track)
     sh->codec->extradata = track->private_data;
     sh->codec->extradata_size = track->private_size;
 
+    if (!strcmp(sh->codec->codec, "arib_caption")) {
+        struct AVCodecParameters **lav = talloc_ptrtype(track, lav);
+        if (!lav)
```
Comment (jeeb) [inline demux/demux_mkv.c:1874]: `talloc_ptrtype -> ta_xnew_ptrtype -> ta_oom_g(ta_new_ptrtype(__VA_ARGS__)) `
so I think unless the ta_ stuff somehow gets overridden by something this check should never fail.

File: demux/demux_mkv.c [hunk 3]
```diff
@@ -1862,6 +1869,45 @@ static int demux_mkv_open_sub(demuxer_t *demuxer, mkv_track_t *track)
     sh->codec->extradata = track->private_data;
     sh->codec->extradata_size = track->private_size;
 
+    if (!strcmp(sh->codec->codec, "arib_caption")) {
```
Comment (jeeb) [inline demux/demux_mkv.c:1872]: Right after this if, you should probably check if `private_size` contains at least three bytes.

File: demux/demux_mkv.c [hunk 4]
```diff
@@ -1862,6 +1869,45 @@ static int demux_mkv_open_sub(demuxer_t *demuxer, mkv_track_t *track)
...
+        struct AVCodecParameters **lav = talloc_ptrtype(track, lav);
+        if (!lav)
+            return 1;
+
+        talloc_set_destructor(lav, avcodec_par_destructor);
+
+        *lav = sh->codec->lav_codecpar = avcodec_parameters_alloc();
+        if (!sh->codec->lav_codecpar)
```
Comment (jeeb) [inline demux/demux_mkv.c:1880]: If we follow the classic mpv OOM handling style, probably `MP_HANDLE_OOM(*lav)`.

I would also recommend taking a pointer towards the codecpar to not have to have `sh->codec->lav_codecpar->` everywhere as it's poked in multiple spots?

File: demux/demux_mkv.c [hunk 5]
```diff
@@ -1862,6 +1869,45 @@ static int demux_mkv_open_sub(demuxer_t *demuxer, mkv_track_t *track)
...
+                sh->codec->lav_codecpar->profile = FF_PROFILE_ARIB_PROFILE_C;
+            break;
+        }
+        if (sh->codec->lav_codecpar->profile == FF_PROFILE_UNKNOWN)
+            MP_WARN(demuxer, "ARIB caption profile %02x / %04x not supported.\n",
+                    component_tag, data_component_id);
+
+        talloc_free(track->private_data);
+        track->private_data = NULL;
+        track->private_size = 0;
```
Comment (jeeb) [inline demux/demux_mkv.c:1908]: I don't see this in the audio path? I know the muxers generally don't utilize this data, but why free it? Especially if it was just set to `sh->codec->extradata = track->private_data` before the codec name check, and thus that side might become an invalid pointer if it's just freed here?

**Review by jeeb — CHANGES_REQUESTED** [review]
The actual component tag and data component id checks are pretty much what I wrote in FFmpeg's MPEG-TS demuxer, so in general this looks LGTM.

**Comment by rcombs** [comment]
Applied all suggested changes.

**Review by jeeb — APPROVED** [review]

**Pull request merged**